### DISCUSSION TITLE
Code Anslysis with disbaled live analysis does not show all warnings.

### DISCUSSION THREAD

**Discussion by LukasSchwendemann** [dotnet/roslyn#41365] [General]
**Version Used**: 
dotnet core 3.1
Visual Studio 2019 (16.4.3)
Microsoft.CodeAnalysis.FxCopAnalyzers 2.6.4
StyleCop.Analyzers 1.1.118
SonarAnalyzer.CSharp 8.3.0.14607

**Steps to Reproduce**:

1. Enable Code analysis with "run on build" and "run on live analysis"
2. Using a global rule set that defies all the rules that should be evaluated
3. using project specific rule sets referencing the global rule set
4. All expected warnings are shown
![image](https://user-images.githubusercontent.com/15869411/73666213-babfce00-46a2-11ea-9dfc-6ba119f1d13d.png)
5. The installed analyzers are shown in the team explorer in the projects under dependencies.
![image](https://user-images.githubusercontent.com/15869411/73666278-d3c87f00-46a2-11ea-9dbe-6467282ada47.png)
5. Disable option "run on live analysis" for Code Analysis
6. Rebuild entire solution.

**Expected Behavior**:
- The installed analyzers are still shown in the team explorer under dependencies
- the same rules are shown as with enabled "run on live analysis"
- running dotnet build after dotnet clean should show all the warnings in console as Visual Studio

**Actual Behavior**:
- many warnings are not shown anymore
![image](https://user-images.githubusercontent.com/15869411/73666465-24d87300-46a3-11ea-8e2c-311ba4c5eac6.png)
- the installed analyzers are not shown anymore
![image](https://user-images.githubusercontent.com/15869411/73666484-2e61db00-46a3-11ea-82e6-47f4cd43504a.png)
- running dotnet build after dotnet clean does not show the warnings in the console even if live analysis is enabled.

**Comment by mavasani** [comment]
> many warnings are not shown anymore

This is by-design. By disabling "Run on live analysis" you explicitly requested not to compute and/or continuously analyze the source code to show current live analysis diagnostics. Explicit build is done against a specific snapshot, and you should switch the "Build + Intellisense" error list combo box to "Build only" to see warnings reported on last build.

> the installed analyzers are not shown anymore

I believe this should be fixed in latest 16.5 Preview drop, can you please try?

> running dotnet build after dotnet clean should show all the warnings in console as Visual Studio

Are you suggesting that doing an explicit build in VS shows more warnings then doing so in console?

**Comment by LukasSchwendemann** [comment]
What do you mean with

> Explicit build is done against a specific snapshot.

What my understanding of the feature "Run on live analysis" is. Is that it shows me ALL the warnings the analyzer finds while I'm producing new code and the already existing ones as well. If I'm disabling this the warnings will not show up anymore for the intelliSense analysis.

But if I Rebuild my entire Solution it should just produce the same output as I would use the "Run on live analysis". So directly after rebuilding the result of "Build + Intellisense" and "Build only" should be the same. Or Did I get something wrong? In my case it just does not produce the same result, there are plenty of warnings missing in the "Build only" selection.

To draw my scenario further. I just want to see all my warnings in the output of my CI-Build and this is not the case.

These are the outputs after a complete rebuild with enabled Live Analysis:
![image](https://user-images.githubusercontent.com/15869411/73682256-51e74e80-46c0-11ea-9de5-933a4eb65840.png)
![image](https://user-images.githubusercontent.com/15869411/73682279-5e6ba700-46c0-11ea-9dc9-df025f30af0e.png)
![image](https://user-images.githubusercontent.com/15869411/73682295-675c7880-46c0-11ea-8422-96d84cfe95aa.png)
Is this normal?

> Are you suggesting that doing an explicit build in VS shows more warnings then doing so in console?

No, I want them to show the same but all.

**Comment by mavasani** [comment]
> What my understanding of the feature "Run on live analysis" is. Is that it shows me ALL the warnings the analyzer finds while I'm producing new code and the already existing ones as well. If I'm disabling this the warnings will not show up anymore for the intelliSense analysis.
But if I Rebuild my entire Solution it should just produce the same output as I would use the "Run on live analysis". 

There seems to be a slight misunderstanding on the feature, and I think I understand the cause for confusion. One might assume that live analysis is only additive, i.e. it can only add new diagnostics on top of the ones from last build. That is however not true. If build produced some diagnostics, and you changed the sources that fix some of these diagnostics, we would still need to re-trigger live analysis to identify the new set of applicable diagnostics for the current source snapshot, otherwise you can have stale diagnostics. However, retriggering analysis this will violate the user preference of not doing any live analysis and also consume resources. I think you have 2 alternatives when you disable live analysis:

1. Only use the "Build Only" combo box option to get an affirmative list of analyzer diagnostics from last time you invoked a build. Live analysis (and hence "Build + Intellisense" setting) will only show up-to-date compiler diagnostics.
2. In VS2019 16.5, we repurposed the "Run Code Analysis" command under "Analyze" top level menu to force execute/complete live analysis for selected project or solution (see https://github.com/dotnet/roslyn/pull/39596). You can use this command to force complete all the analyzers and populate the error list with these analyzer diagnostics. Note that the analyzer diagnostics will not automatically refresh if you edit sources after executing this command, as no live analysis is executing. You will need to invoke the command again to get the current set of analyzer diagnostics.

**Comment by LukasSchwendemann** [comment]
Thank you for the support! I had some misunderstanding regarding the usage of the "build" and "build + intelligent" selection. But I still don't think we are talking on the same thing. Because Solution 1 will not do anything for me. because the build / compiler does not report all the warnings that are in the code. I try to describe it from another angle:

1. enable the live analysis in all the projects.
2. clean my solution via "dotnet clean" in cmd.
3. rebuild the solution via "dotnet build" in cmd.
4. there are 26 Warnings reported by the build
5. looking into Visual Studio using the "IntelleSense only" selection I see 836 warnings there are Issues of the rule SA1613, SA1623, SA1629 from the stylecop analyzer. S1118, S1135, S4018 of the sonar analyzer that are completely not reported by the build. I see warnings in "Intellisense Only" view that are never reported by the build. Is that intended?

The important thing for me is to have all the existing warnings in the code reported by the compiler to use created reports in my CI-build on Azure DevOps.

**Comment by mavasani** [comment]
@LukasSchwendemann So, your primary issue is that after installing StyleCop analyzers and Sonar analyzers, you are seeing warnings from intellisense, but these are not showing up in command line build, correct? If so, that is not expected behavior. Can you provide a small standalone repro for us to investigate?
